File: internal/model/website.go
package model

import "time"

// 站点类型。本期实现 static 与 proxy，其余类型待运行环境模块落地后开放。
const (
	SiteTypeStatic = "static"
	SiteTypeProxy  = "proxy"
	SiteTypePHP    = "php"
)

// 站点状态机（docs/07 7.2.3）。
const (
	SiteStatusRunning     = "running"
	SiteStatusStopped     = "stopped"
	SiteStatusError       = "error"
	SiteStatusDeploying   = "deploying"
	SiteStatusMaintenance = "maintenance"
)

// Web 服务引擎。
const (
	ServerTypeNginx     = "nginx"
	ServerTypeOpenResty = "openresty"
)

// 配置类型与来源。
const (
	SiteConfigTypeNginx      = "nginx"
	SiteConfigSourcePanel    = "panel"
	SiteConfigSourceManual   = "manual_edit"
	SiteConfigSourceRollback = "rollback"
)

// 配置下发状态。
const (
	DeployStatusPending    = "pending"
	DeployStatusDeployed   = "deployed"
	DeployStatusFailed     = "failed"
	DeployStatusRolledBack = "rolled_back"
)

// WebSite 对应 web_site 表，是站点聚合根。name 同时作为 vhost 文件名与默认目录名。
type WebSite struct {
	Base
	SoftDelete
	NodeID           int64      `gorm:"column:node_id;not null;default:0"`
	Name             string     `gorm:"column:name;not null"`
	Type             string     `gorm:"column:type;not null;default:'static'"`
	ServerType       string     `gorm:"column:server_type;not null;default:'nginx'"`
	RootPath         string     `gorm:"column:root_path;not null;default:''"`
	IndexFiles       string     `gorm:"column:index_files;not null;default:'index.html index.htm'"`
	RuntimeID        int64      `gorm:"column:runtime_id;not null;default:0"`
	PHPVersion       string     `gorm:"column:php_version;not null;default:''"`
	ProxyTarget      string     `gorm:"column:proxy_target;not null;default:''"`
	ProxyHost        string     `gorm:"column:proxy_host;not null;default:''"`
	ProxyWebSocket   bool       `gorm:"column:proxy_websocket;not null;default:true"`
	ListenPort       int        `gorm:"column:listen_port;not null;default:80"`
	SSLPort          int        `gorm:"column:ssl_port;not null;default:443"`
	SSLEnabled       bool       `gorm:"column:ssl_enabled;not null;default:false"`
	CertID           int64      `gorm:"column:cert_id;not null;default:0"`
	ForceHTTPS       bool       `gorm:"column:force_https;not null;default:false"`
	HTTP2Enabled     bool       `gorm:"column:http2_enabled;not null;default:true"`
	HTTP3Enabled     bool       `gorm:"column:http3_enabled;not null;default:false"`
	WAFEnabled       bool       `gorm:"column:waf_enabled;not null;default:false"`
	AccessLogEnabled bool       `gorm:"column:access_log_enabled;not null;default:true"`
	ErrorLogEnabled  bool       `gorm:"column:error_log_enabled;not null;default:true"`
	AccessLogPath    string     `gorm:"column:access_log_path;not null;default:''"`
	ErrorLogPath     string     `gorm:"column:error_log_path;not null;default:''"`
	RateLimitJSON    string     `gorm:"column:rate_limit_json"`
	SettingsJSON     string     `gorm:"column:settings_json"`
	Status           string     `gorm:"column:status;not null;default:'running'"`
	ConfigVersion    int        `gorm:"column:config_version;not null;default:1"`
	ExpireAt         *time.Time `gorm:"column:expire_at"`
	GroupName        string     `gorm:"column:group_name;not null;default:'default'"`
	Remark           string     `gorm:"column:remark;not null;default:''"`
}

// TableName 指定表名。
func (WebSite) TableName() string { return "web_site" }

// WebSiteDomain 对应 web_site_domain 表，一行是一个「域名 + 端口」绑定。
type WebSiteDomain struct {
	Base
	SoftDelete
	SiteID         int64      `gorm:"column:site_id;not null"`
	Domain         string     `gorm:"column:domain;not null"`
	Port           int        `gorm:"column:port;not null;default:80"`
	IsPrimary      bool       `gorm:"column:is_primary;not null;default:false"`
	IsWildcard     bool       `gorm:"column:is_wildcard;not null;default:false"`
	CertID         int64      `gorm:"column:cert_id;not null;default:0"`
	DNSResolvedIP  string     `gorm:"column:dns_resolved_ip;not null;default:''"`
	DNSCheckAt     *time.Time `gorm:"column:dns_check_at"`
	DNSCheckResult string     `gorm:"column:dns_check_result;not null;default:''"`
}

// TableName 指定表名。
func (WebSiteDomain) TableName() string { return "web_site_domain" }

// WebSiteConfig 对应 web_site_config 表，是只增不改的配置版本快照。
type WebSiteConfig struct {
	Base
	SoftDelete
	SiteID        int64      `gorm:"column:site_id;not null"`
	Version       int        `gorm:"column:version;not null"`
	ConfigType    string     `gorm:"column:config_type;not null;default:'nginx'"`
	Content       string     `gorm:"column:content;not null"`
	ContentHash   string     `gorm:"column:content_hash;not null"`
	Source        string     `gorm:"column:source;not null;default:'panel'"`
	IsCurrent     bool       `gorm:"column:is_current;not null;default:false"`
	DeployStatus  string     `gorm:"column:deploy_status;not null;default:'pending'"`
	DeployAt      *time.Time `gorm:"column:deploy_at"`
	DeployError   string     `gorm:"column:deploy_error;not null;default:''"`
	ChangeSummary string     `gorm:"column:change_summary;not null;default:''"`
}

// TableName 指定表名。
func (WebSiteConfig) TableName() string { return "web_site_config" }

File: web/src/api/website.ts
import { request } from '@/utils/request'

/* 类型与后端 internal/model/dto_website.go、internal/service/site 保持一致。 */

export type SiteType = 'static' | 'proxy'
export type SiteStatus = 'running' | 'stopped' | 'error' | 'deploying' | 'maintenance'
export type DeployStatus = 'pending' | 'deployed' | 'failed' | ''

export interface SiteDomain {
  /** 雪花 ID 以字符串传输，避开 JS 整数精度限制；创建时不传 */
  id?: string
  domain: string
  port: number
  isPrimary?: boolean
  isWildcard?: boolean
}

export interface SiteSummary {
  id: string
  name: string
  type: SiteType
  status: SiteStatus
  serverType: string
  primaryDomain: string
  domains: SiteDomain[]
  rootPath: string
  proxyTarget: string
  sslEnabled: boolean
  groupName: string
  remark: string
  configVersion: number
  deployStatus: DeployStatus
  deployError: string
  /** Unix 毫秒；0 表示永不过期 */
  expireAt: number
  createdAt: number
  updatedAt: number
}

export interface SiteDetail extends SiteSummary {
  indexFiles: string[]
  accessLogEnabled: boolean
  errorLogEnabled: boolean
  accessLogPath: string
  errorLogPath: string
  proxyHost: string
  proxyWebSocket: boolean
  vhostFile: string
  rootExists: boolean
}

export interface SiteListResult {
  items: SiteSummary[]
  total: number
  page: number
  size: number
}

/** 建站前提：宿主机 Web 服务状态与真实支持的站点类型。 */
export interface WebsiteEnv {
  installed: boolean
  kind: string
  bin: string
  version: string
  vhostDir: string
  /** null 表示 nginx -T 不可用，无法判断主配置是否 include 面板目录 */
  vhostIncluded: boolean | null
  configOK: boolean
  message: string
  wwwRoot: string
  logDir: string
  supportedTypes: SiteType[]
  siteCount: number
}

export interface CreateSitePayload {
  name: string
  type: SiteType
  domains: SiteDomain[]
  root?: string
  indexFiles?: string[]
  remark?: string
  groupName?: string
  expireAt?: number
  proxyTarget?: string
  proxyHost?: string
  proxyWebSocket?: boolean
  createDefaultPage?: boolean
  accessLogEnabled?: boolean
  errorLogEnabled?: boolean
}

export interface UpdateSitePayload {
  domains?: SiteDomain[]
  root?: string
  indexFiles?: string[]
  remark?: string
  groupName?: string
  expireAt?: number
  proxyTarget?: string
  proxyHost?: string
  proxyWebSocket?: boolean
  accessLogEnabled?: boolean
  errorLogEnabled?: boolean
}

export interface SiteConfigVersion {
  version: number
  source: string
  isCurrent: boolean
  deployStatus: DeployStatus
  deployError: string
  contentHash: string
  changeSummary: string
  createdAt: number
  deployAt: number
}

export interface SiteConfig {
  siteId: string
  name: string
  file: string
  version: number
  content: string
  versions: SiteConfigVersion[]
}

// 网站模块的失败提示由页面自己给出（列表用告警区、详情用 Message），
// 因此统一带 silent，避免请求层再弹一条重复的全局提示。
export function fetchWebsiteEnv() {
  return request<WebsiteEnv>({ url: '/website/env', method: 'get', silent: true })
}

export function fetchSites(params: { page?: number; pageSize?: number; keyword?: string }) {
  return request<SiteListResult>({ url: '/website/sites', method: 'get', params, silent: true })
}

export function fetchSite(id: string) {
  return request<SiteDetail>({ url: `/website/sites/${id}`, method: 'get', silent: true })
}

export function createSite(payload: CreateSitePayload) {
  return request<SiteDetail>({ url: '/website/sites', method: 'post', data: payload, silent: true })
}

export function updateSite(id: string, payload: UpdateSitePayload) {
  return request<SiteDetail>({
    url: `/website/sites/${id}`,
    method: 'put',
    data: payload,
    silent: true,
  })
}

export function changeSiteStatus(id: string, target: Extract<SiteStatus, 'running' | 'stopped'>) {
  return request<SiteDetail>({
    url: `/website/sites/${id}/status`,
    method: 'post',
    data: { target },
    silent: true,
  })
}

export function rebuildSite(id: string) {
  return request<SiteDetail>({ url: `/website/sites/${id}/rebuild`, method: 'post', silent: true })
}

export function deleteSite(id: string, deleteRoot: boolean) {
  return request<{ deleted: boolean; rootRecycled: boolean }>({
    url: `/website/sites/${id}`,
    method: 'delete',
    params: { deleteRoot: deleteRoot ? 'true' : 'false' },
    silent: true,
  })
}

export function fetchSiteConfig(id: string) {
  return request<SiteConfig>({ url: `/website/sites/${id}/config`, method: 'get', silent: true })
}

export function saveSiteConfig(id: string, content: string, version: number) {
  return request<SiteConfig>({
    url: `/website/sites/${id}/config`,
    method: 'put',
    data: { content, version },
    silent: true,
  })
}

export function rollbackSiteConfig(id: string, version: number) {
  return request<SiteConfig>({
    url: `/website/sites/${id}/config/rollback/${version}`,
    method: 'post',
    silent: true,
  })
}

export function reloadWebServer() {
  return request<{ reloaded: boolean }>({
    url: '/website/nginx/reload',
    method: 'post',
    silent: true,
  })
}

export function testWebServer() {
  return request<{ ok: boolean; output: string }>({
    url: '/website/nginx/test',
    method: 'post',
    silent: true,
  })
}

/** 静态缓存档位，与后端 model.CacheProfile* 一致。 */
export type CacheProfile = 'none' | 'hour' | 'day' | 'month' | 'year'

export interface AntiLeechSetting {
  enabled: boolean
  extensions: string[]
  allowedReferers: string[]
  allowEmptyReferer: boolean
  returnCode: number
}

export interface AccessControlSetting {
  enabled: boolean
  allowIps: string[]
  denyIps: string[]
  denyUserAgents: string[]
}

export interface RedirectRule {
  enabled: boolean
  path: string
  target: string
  code: number
  keepPath: boolean
  keepQuery: boolean
  remark: string
}

export interface RateLimitSetting {
  enabled: boolean
  bandwidthKb: number
  burstKb: number
}

export interface SiteSettings {
  siteId: string
  name: string
  type: SiteType
  indexFiles: string[]
  runPath: string
  autoIndex: boolean
  cacheProfile: CacheProfile
  antiLeech: AntiLeechSetting
  accessControl: AccessControlSetting
  redirects: RedirectRule[]
  rateLimit: RateLimitSetting
  accessLogEnabled: boolean
  errorLogEnabled: boolean
  accessLogPath: string
  errorLogPath: string
  rootPath: string
  /** 后端给出的可选缓存档位，前端不写死枚举 */
  cacheProfiles: CacheProfile[]
}

/** 保存为整体替换语义：一次提交完整状态，避免部分更新的歧义。 */
export type SaveSiteSettingsPayload = Omit<
  SiteSettings,
  'siteId' | 'name' | 'type' | 'accessLogPath' | 'errorLogPath' | 'rootPath' | 'cacheProfiles'
>

export interface SiteRewrite {
  siteId: string
  name: string
  /** 内置模板代码、custom 或 none */
  templateCode: string
  content: string
  enabled: boolean
  /** 片段落盘路径，便于用户核对 */
  file: string
  updatedAt: number
}

export interface SiteRewriteTemplate {
  code: string
  name: string
  content: string
}

export function fetchSiteSettings(id: string) {
  return request<SiteSettings>({ url: `/website/sites/${id}/settings`, method: 'get', silent: true })
}

export function saveSiteSettings(id: string, payload: SaveSiteSettingsPayload) {
  return request<SiteSettings>({
    url: `/website/sites/${id}/settings`,
    method: 'put',
    data: payload,
    silent: true,
  })
}

export function fetchRewriteTemplates() {
  return request<{ items: SiteRewriteTemplate[] }>({
    url: '/website/rewrite/templates',
    method: 'get',
    silent: true,
  })
}

export function fetchSiteRewrite(id: string) {
  return request<SiteRewrite>({ url: `/website/sites/${id}/rewrite`, method: 'get', silent: true })
}

export function saveSiteRewrite(
  id: string,
  payload: { templateCode: string; content: string; enabled: boolean },
) {
  return request<SiteRewrite>({
    url: `/website/sites/${id}/rewrite`,
    method: 'put',
    data: payload,
    silent: true,
  })
}

export type SiteLogType = 'access' | 'error'

export interface SiteLog {
  siteId: string
  name: string
  type: SiteLogType
  path: string
  /** 站点设置里的日志开关；关闭时文件通常不存在 */
  enabled: boolean
  exists: boolean
  size: number
  modifiedAt: number
  lines: string[]
  /** 真表示只返回了尾部，文件比窗口更长 */
  truncated: boolean
}

export function fetchSiteLog(id: string, type: SiteLogType, lines: number) {
  return request<SiteLog>({
    url: `/website/sites/${id}/logs`,
    method: 'get',
    params: { type, lines },
    silent: true,
  })
}

export function clearSiteLog(id: string, type: SiteLogType) {
  return request<SiteLog>({
    url: `/website/sites/${id}/logs/clear`,
    method: 'post',
    params: { type },
    silent: true,
  })
}
